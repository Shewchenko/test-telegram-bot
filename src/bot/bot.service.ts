import { Injectable } from '@nestjs/common';
process.env.NTBA_FIX_319 = String(1);
import * as TelegramBot from 'node-telegram-bot-api';
import { ConfigService } from '@nestjs/config';

import { InlineBuilder } from './inline.builder';

@Injectable()
export class BotService {
  private readonly _bot: TelegramBot;

  constructor(private readonly configService: ConfigService) {
    this._bot = new TelegramBot(configService.get<string>('BOT_TOKEN'), {
      polling: true,
    });
  }

  get bot(): TelegramBot {
    return this._bot;
  }

  async editMsg(
    txt: string,
    msg: TelegramBot.Message,
    btns: InlineBuilder,
    options?: TelegramBot.EditMessageTextOptions,
  ): Promise<TelegramBot.Message | boolean> {
    let opts: TelegramBot.EditMessageTextOptions = {
      chat_id: msg.chat.id,
      message_id: msg.message_id,
      reply_markup: btns.getReplyEdit(),
      parse_mode: 'Markdown',
    };
    if (options) {
      opts = { ...opts, ...options };
    }
    return this.bot.editMessageText(txt, opts);
  }

  async editMessageReplyMarkup(
    replyMarkup: TelegramBot.InlineKeyboardMarkup,
    options?: TelegramBot.EditMessageReplyMarkupOptions,
  ): Promise<TelegramBot.Message | boolean> {
    return this._bot.editMessageReplyMarkup(replyMarkup, options);
  }

  async sendMessage(
    msg: TelegramBot.Message,
    txt: string,
    options?: TelegramBot.SendMessageOptions,
  ): Promise<TelegramBot.Message> {
    return this.bot.sendMessage(msg.chat.id, txt, options);
  }
}
