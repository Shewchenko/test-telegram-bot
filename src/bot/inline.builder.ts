import TelegramBot, {
  InlineKeyboardButton,
  InlineKeyboardMarkup,
} from 'node-telegram-bot-api';

export class InlineBuilder {
  protected lines: InlineKeyboardButton[][] = [[]];

  constructor();
  constructor(params: InlineKeyboardButton);
  constructor(params: InlineKeyboardButton[]);
  constructor(
    params: InlineKeyboardButton | InlineKeyboardButton[] = undefined,
  ) {
    if (typeof params === 'undefined') {
      return;
    }
    if (!Array.isArray(params)) {
      params = [params];
    }
    params.forEach((p) => this.addButton(p));
  }

  addButton(params: InlineKeyboardButton): this {
    this.lines[this.lines.length - 1].push(params);
    return this;
  }

  nextLine() {
    this.lines.push([]);
    return this;
  }

  mergeLines(lines: []) {
    this.lines.push(lines);
  }

  getLines(): InlineKeyboardButton[][] {
    return this.lines;
  }

  getReply(): TelegramBot.SendMessageOptions {
    return <TelegramBot.SendMessageOptions>{
      reply_markup: this.getReplyEdit(),
    };
  }

  getReplyEdit(): InlineKeyboardMarkup {
    return <InlineKeyboardMarkup>{
      inline_keyboard: this.lines,
    };
  }
}
