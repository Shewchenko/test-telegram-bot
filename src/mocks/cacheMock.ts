/* istanbul ignore file */
// tslint:disable
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck

import { Cache, CachingConfig, Store, WrapArgsType } from 'cache-manager';

export class CacheMock implements Cache {
  store: Store;
  async del(key: string, callback: (error: any) => void): Promise<void>;
  async del(key: string): Promise<any>;
  async del(
    key: string,
    callback?: (error: any) => void,
  ): Promise<void | Promise<any>> {
    return undefined;
  }

  async get<T>(
    key: string,
    callback: (error: any, result: T | undefined) => void,
  ): Promise<void>;
  async get<T>(key: string): Promise<T | undefined>;
  async get(key: string, callback?): Promise<any> {}

  async reset(): Promise<void>;
  async reset(cb: () => void): Promise<void>;
  async reset(cb?: () => void): Promise<Promise<void> | void> {
    return undefined;
  }

  async set<T>(key: string, value: T, options?: CachingConfig): Promise<T>;
  async set<T>(key: string, value: T, ttl: number): Promise<T>;
  async set<T>(
    key: string,
    value: T,
    options: CachingConfig,
    callback: (error: any) => void,
  ): Promise<void>;
  async set<T>(
    key: string,
    value: T,
    ttl: number,
    callback: (error: any) => void,
  ): Promise<void>;
  async set(
    key: string,
    value,
    options?: CachingConfig | number,
    callback?: (error: any) => void,
  ): Promise<any> {}

  wrap<T>(...args: WrapArgsType<T>[]): Promise<T> {
    return Promise.resolve(undefined);
  }
}
