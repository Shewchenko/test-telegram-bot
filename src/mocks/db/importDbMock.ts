import { TypeOrmModule } from '@nestjs/typeorm';

import { RegisterEntityMocks } from '../registerEntityMocks';

export class ImportDbMock {
  public static get() {
    return [
      TypeOrmModule.forRoot({
        type: 'sqlite',
        database: ':memory:',
        entities: RegisterEntityMocks.getEntityMocks(),
        synchronize: true,
      }),
      TypeOrmModule.forFeature(RegisterEntityMocks.getEntityMocks()),
    ];
  }
}
