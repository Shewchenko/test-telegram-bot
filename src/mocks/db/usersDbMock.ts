import { IUser } from '../../users/contracts/iUser';

export class UsersDbMock {
  public static get(): IUser[] {
    return [
      {
        name: 'user1',
        tgId: 1,
      },
      {
        name: 'user2',
        tgId: 2,
      },
    ] as IUser[];
  }
}
