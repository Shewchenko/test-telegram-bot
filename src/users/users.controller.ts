import { Controller, Post, Body, BadRequestException } from '@nestjs/common';

import { UsersService } from './users.service';
import { UpdateMessageDto } from './dto/updateMessageDto';

@Controller()
export class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @Post('update')
  async updateMessage(@Body() updateMessageDto: UpdateMessageDto) {
    try {
      return await this.usersService.updateListBtns(updateMessageDto.list);
    } catch (e) {
      throw new BadRequestException(e.message);
    }
  }
}
