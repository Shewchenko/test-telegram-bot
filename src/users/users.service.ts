import { CACHE_MANAGER, Inject, Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import TelegramBot from 'node-telegram-bot-api';
import { Cache } from 'cache-manager';

import { IUser } from './contracts/iUser';
import { User } from './entities/user.entity';
import { BotService } from '../bot/bot.service';
import { InlineBuilder } from '../bot/inline.builder';
import { Tools } from '../utils/tools';
import { TBotCbUser } from './types/tBotCbUser';

@Injectable()
export class UsersService {
  private readonly logger = new Logger(UsersService.name);

  private static USER_KEY = 'userKey';
  private static UPDATE_DATA_KEY = 'updateData';
  private static LIST_LENGTH = 8;

  private static DEFAULT_BTNS = [
    'Btn1',
    'Btn2',
    'Btn3',
    'Btn4',
    'Btn5',
    'Btn6',
    'Btn7',
    'Btn8',
  ];

  constructor(
    @InjectRepository(User)
    private readonly repository: Repository<IUser>,
    private readonly botService: BotService,
    @Inject(CACHE_MANAGER)
    private readonly cacheManager: Cache,
  ) {
    this.initBotListening();
  }

  initBotListening() {
    this.botService.bot.onText(/^\/start/, async (msg) => this.start(msg));
    this.botService.bot.onText(/^\/list/, async (msg) =>
      this.sendListBtns(msg),
    );
    this.botService.bot.on('callback_query', async (q) => this.cbQuery(q));
  }

  async cbQuery(query: TelegramBot.CallbackQuery): Promise<void> {
    if (!query?.data) {
      return;
    }
    const params = Tools.parseCbParams<TBotCbUser>(query.data);
    if (params?.t !== 'u') {
      return;
    }
    this.logger.log('Call on TelegramBot.CallbackQuery');

    if (params?.u === 'pag' && params?.page) {
      await this.pagination(params.page);
      return;
    }
  }

  async start(msg: TelegramBot.Message) {
    const model = await this.repository.findOne({ tgId: msg.chat.id });
    if (model) {
      await this.sendListBtns(msg);
      return;
    }
    await this.createUserAndSendListBtns(msg);
  }

  async sendListBtns(msg: TelegramBot.Message): Promise<void> {
    const btns = new InlineBuilder();
    this.buildList(btns, UsersService.DEFAULT_BTNS);
    const res = await this.botService.sendMessage(
      msg,
      `Header`,
      btns.getReply(),
    );
    await this.cacheManager.set(UsersService.USER_KEY, JSON.stringify(res));
  }

  async updateListBtns(arr: string[]): Promise<void> {
    if (!Array.isArray(arr) || !arr.length) {
      throw new Error('Bad update data');
    }
    await this.cacheManager.set(
      UsersService.UPDATE_DATA_KEY,
      JSON.stringify(arr),
    );

    const msg = await this.getMsgByCache();

    const countArr = arr.length;
    arr.splice(UsersService.LIST_LENGTH);
    const btns = new InlineBuilder();
    this.buildList(btns, arr);
    if (countArr > UsersService.LIST_LENGTH) {
      btns.addButton({
        text: `1 / ${Math.ceil(countArr / UsersService.LIST_LENGTH)}`,
        callback_data: 'progress',
      });
      btns.addButton({
        text: `➡`,
        callback_data: Tools.stringifyCbParams<TBotCbUser>({
          t: 'u',
          u: 'pag',
          page: 2,
        }),
      });
    }
    await this.botService.editMessageReplyMarkup(btns.getReplyEdit(), {
      chat_id: msg.chat.id,
      message_id: msg.message_id,
    });
  }

  async getMsgByCache(): Promise<TelegramBot.Message> {
    const user = (await this.cacheManager.get(UsersService.USER_KEY)) as string;
    if (!user) {
      throw new Error('User not found by cache');
    }
    return Tools.parseCbParams<TelegramBot.Message>(user);
  }

  async pagination(page = 1): Promise<void> {
    if (!(typeof page === 'number') || page < 1) {
      this.logger.error('Pagination: invalid param');
      return;
    }
    const msg = await this.getMsgByCache();
    const data = (await this.cacheManager.get(
      UsersService.UPDATE_DATA_KEY,
    )) as string;
    if (!data) {
      throw new Error('Array not found by cache');
    }
    const arr = Tools.parseCbParams<string[]>(data);

    const countArr = arr.length;
    const list = arr.slice(
      UsersService.LIST_LENGTH * (page - 1),
      UsersService.LIST_LENGTH * page,
    );
    const btns = new InlineBuilder();
    this.buildList(btns, list);
    if (countArr > UsersService.LIST_LENGTH) {
      const total = Math.ceil(countArr / UsersService.LIST_LENGTH);
      btns.addButton({
        text: `⬅`,
        callback_data: Tools.stringifyCbParams<TBotCbUser>({
          t: 'u',
          u: 'pag',
          page: page - 1,
        }),
      });
      btns.addButton({
        text: `${page} / ${total}`,
        callback_data: 'progress',
      });
      if (page !== total) {
        btns.addButton({
          text: `➡`,
          callback_data: Tools.stringifyCbParams<TBotCbUser>({
            t: 'u',
            u: 'pag',
            page: page + 1,
          }),
        });
      }
    }
    await this.botService.editMessageReplyMarkup(btns.getReplyEdit(), {
      chat_id: msg.chat.id,
      message_id: msg.message_id,
    });
  }

  async createUserAndSendListBtns(msg: TelegramBot.Message): Promise<void> {
    await this.repository.save(
      this.repository.create({
        tgId: msg.chat.id,
        name: msg.chat.first_name,
      }),
    );
    await this.sendListBtns(msg);
  }

  buildList(btns: InlineBuilder, list: string[]) {
    list.forEach((b, i) => {
      btns.addButton({
        text: b,
        callback_data: Tools.stringifyCbParams<TBotCbUser>({
          t: 'u',
          u: 'cmd',
          d: `btn-${i + 1}`,
        }),
      });
      btns.nextLine();
    });
  }
}
