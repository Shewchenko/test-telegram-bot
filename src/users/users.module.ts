import { CacheModule, Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import * as redisStore from 'cache-manager-redis-store';

import { UsersService } from './users.service';
import { UsersController } from './users.controller';
import { BotModule } from '../bot/bot.module';
import { User } from './entities/user.entity';
import { ConfigModule, ConfigService } from '@nestjs/config';

@Module({
  imports: [
    BotModule,
    TypeOrmModule.forFeature([User]),
    CacheModule.registerAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        store: redisStore,
        host: configService.get('REDIS_HOST'),
        port: configService.get('REDIS_PORT'),
        ttl: configService.get('CACHE_TTL'),
      }),
    }),
  ],
  controllers: [UsersController],
  providers: [UsersService],
})
export class UsersModule {}
