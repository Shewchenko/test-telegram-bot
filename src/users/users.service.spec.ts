import { Test, TestingModule } from '@nestjs/testing';
import { ConfigModule } from '@nestjs/config';
import { Cache } from 'cache-manager';
import TelegramBot from 'node-telegram-bot-api';
import { CACHE_MANAGER } from '@nestjs/common';
import { getConnection } from 'typeorm';

import { UsersService } from './users.service';
import { ImportDbMock } from '../mocks/db/importDbMock';
import { BotService } from '../bot/bot.service';
import { CacheMock } from '../mocks/cacheMock';
import { InlineBuilder } from '../bot/inline.builder';

describe('UsersService', () => {
  let service: UsersService;
  let botService: BotService;
  let cache: Cache;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          load: [
            () => ({
              BOT_TOKEN: 'test',
            }),
          ],
        }),
        ...ImportDbMock.get(),
      ],
      providers: [
        UsersService,
        BotService,
        {
          provide: CACHE_MANAGER,
          useClass: CacheMock,
        },
      ],
    }).compile();

    service = module.get<UsersService>(UsersService);
    botService = module.get<BotService>(BotService);
    cache = module.get<CacheMock>(CACHE_MANAGER);
  });

  afterEach(() => {
    const conn = getConnection();
    return conn.close();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('sendListBtns', () => {
    it('call method', async () => {
      botService.sendMessage = jest.fn();
      cache.set = jest.fn();
      await service.sendListBtns({
        chat: { id: 1 },
        message_id: 10,
      } as TelegramBot.Message);
      expect(botService.sendMessage).toMatchSnapshot();
      expect(cache.set).toMatchSnapshot();
    });
  });

  describe('getMsgByCache', () => {
    it('User not found by cache', async () => {
      expect.assertions(1);
      cache.get = jest.fn().mockReturnValue(undefined);
      try {
        await service.getMsgByCache();
      } catch (e) {
        expect(e.message).toBe('User not found by cache');
      }
    });
    it('success', async () => {
      expect.assertions(1);
      const msg = { chat: { id: 1 }, message_id: 1 } as TelegramBot.Message;
      cache.get = jest.fn().mockReturnValue(JSON.stringify(msg));
      const res = await service.getMsgByCache();
      expect(res).toEqual(msg);
    });
  });

  describe('updateListBtns', () => {
    it('Array is empty', async () => {
      expect.assertions(1);
      try {
        await service.updateListBtns([]);
      } catch (e) {
        expect(e.message).toBe('Bad update data');
      }
    });
    it('List less 9', async () => {
      cache.set = jest.fn();
      cache.get = jest
        .fn()
        .mockReturnValue(`{"chat": { "id": 1 }, "message_id": 1}`);
      botService.editMessageReplyMarkup = jest.fn();
      await service.updateListBtns(['1', '2']);
      expect(botService.editMessageReplyMarkup).toMatchSnapshot();
    });
    it('List more 9', async () => {
      cache.set = jest.fn();
      cache.get = jest
        .fn()
        .mockReturnValue(`{"chat": { "id": 1 }, "message_id": 1}`);
      botService.editMessageReplyMarkup = jest.fn();
      await service.updateListBtns([
        'btn',
        'btn',
        'btn',
        'btn',
        'btn',
        'btn',
        'btn',
        'btn',
        'btn',
        'btn',
      ]);
      expect(botService.editMessageReplyMarkup).toMatchSnapshot();
    });
  });

  describe('buildList', () => {
    it.each([
      [['name1', 'name2', 'name3', 'name4']],
      [['name1', 'name2', 'name3']],
      [['name1', 'name2']],
      [['name1']],
    ])(`List: %s`, async (arr) => {
      const btns = new InlineBuilder();
      service.buildList(btns, arr);
      expect(btns.getLines()).toMatchSnapshot();
    });
  });
});
