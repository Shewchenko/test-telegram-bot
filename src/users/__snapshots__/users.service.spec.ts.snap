// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersService buildList List: [ 'name1' ] 1`] = `
Array [
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-1\\"}",
      "text": "name1",
    },
  ],
  Array [],
]
`;

exports[`UsersService buildList List: [ 'name1', 'name2' ] 1`] = `
Array [
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-1\\"}",
      "text": "name1",
    },
  ],
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-2\\"}",
      "text": "name2",
    },
  ],
  Array [],
]
`;

exports[`UsersService buildList List: [ 'name1', 'name2', 'name3' ] 1`] = `
Array [
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-1\\"}",
      "text": "name1",
    },
  ],
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-2\\"}",
      "text": "name2",
    },
  ],
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-3\\"}",
      "text": "name3",
    },
  ],
  Array [],
]
`;

exports[`UsersService buildList List: [ 'name1', 'name2', 'name3', 'name4' ] 1`] = `
Array [
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-1\\"}",
      "text": "name1",
    },
  ],
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-2\\"}",
      "text": "name2",
    },
  ],
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-3\\"}",
      "text": "name3",
    },
  ],
  Array [
    Object {
      "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-4\\"}",
      "text": "name4",
    },
  ],
  Array [],
]
`;

exports[`UsersService sendListBtns call method 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "chat": Object {
          "id": 1,
        },
        "message_id": 10,
      },
      "Header",
      Object {
        "reply_markup": Object {
          "inline_keyboard": Array [
            Array [
              Object {
                "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-1\\"}",
                "text": "Btn1",
              },
            ],
            Array [
              Object {
                "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-2\\"}",
                "text": "Btn2",
              },
            ],
            Array [
              Object {
                "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-3\\"}",
                "text": "Btn3",
              },
            ],
            Array [
              Object {
                "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-4\\"}",
                "text": "Btn4",
              },
            ],
            Array [
              Object {
                "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-5\\"}",
                "text": "Btn5",
              },
            ],
            Array [
              Object {
                "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-6\\"}",
                "text": "Btn6",
              },
            ],
            Array [
              Object {
                "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-7\\"}",
                "text": "Btn7",
              },
            ],
            Array [
              Object {
                "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-8\\"}",
                "text": "Btn8",
              },
            ],
            Array [],
          ],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`UsersService sendListBtns call method 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "userKey",
      "{\\"chat\\":{\\"id\\":1},\\"message_id\\":10}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`UsersService updateListBtns List less 9 1`] = `[MockFunction]`;

exports[`UsersService updateListBtns List more 9 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "inline_keyboard": Array [
          Array [
            Object {
              "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-1\\"}",
              "text": "btn",
            },
          ],
          Array [
            Object {
              "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-2\\"}",
              "text": "btn",
            },
          ],
          Array [
            Object {
              "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-3\\"}",
              "text": "btn",
            },
          ],
          Array [
            Object {
              "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-4\\"}",
              "text": "btn",
            },
          ],
          Array [
            Object {
              "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-5\\"}",
              "text": "btn",
            },
          ],
          Array [
            Object {
              "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-6\\"}",
              "text": "btn",
            },
          ],
          Array [
            Object {
              "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-7\\"}",
              "text": "btn",
            },
          ],
          Array [
            Object {
              "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"cmd\\",\\"d\\":\\"btn-8\\"}",
              "text": "btn",
            },
          ],
          Array [
            Object {
              "text": "1 / 2",
            },
            Object {
              "callback_data": "{\\"t\\":\\"u\\",\\"u\\":\\"pag\\",\\"page\\":2}",
              "text": "âž¡",
            },
          ],
        ],
      },
      Object {
        "chat_id": 1,
        "message_id": 1,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
